RawSpeed/LJpegPlain.cpp

```diff
@@ -62,493 +49,127 @@ void LJpegPlain::decodeScan() {
-  _ASSERTE(frame.compInfo[0].superH == 2);   // Check if this is a valid state
-  _ASSERTE(frame.compInfo[0].superV == 2);   // Check if this is a valid state
-  _ASSERTE(frame.compInfo[1].superH == 1);   // Check if this is a valid state
-  _ASSERTE(frame.compInfo[1].superV == 1);   // Check if this is a valid state
-  _ASSERTE(frame.compInfo[2].superH == 1);   // Check if this is a valid state
-  _ASSERTE(frame.compInfo[2].superV == 1);   // Check if this is a valid state
-  _ASSERTE(frame.cps == COMPS);
-  _ASSERTE(skipX == 0);
-
-  HuffmanTable *dctbl1 = huff[frame.compInfo[0].dcTblNo];
-  HuffmanTable *dctbl2 = huff[frame.compInfo[1].dcTblNo];
-  HuffmanTable *dctbl3 = huff[frame.compInfo[2].dcTblNo];
-
-  ushort16 *predict;      // Prediction pointer
-
-  mRaw->metadata.subsampling.x = 2;
-  mRaw->metadata.subsampling.y = 2;
-
-  uchar8 *draw = mRaw->getData();
-  // Fix for Canon 6D mRaw, which has flipped width & height
-  uint32 real_h = mCanonFlipDim ? frame.w : frame.h;
+  _ASSERTE(!(slicesW.size() > 1 && skipX));
+  _ASSERTE(frame.compInfo[0].superH == X_S_F);
+  _ASSERTE(frame.compInfo[0].superV == Y_S_F);
+  _ASSERTE(frame.compInfo[1].superH == 1);
+  _ASSERTE(frame.compInfo[1].superV == 1);
+  _ASSERTE(frame.cps == N_COMP);
+  _ASSERTE(skipX == 0 || X_S_F == 1);
+
+  // old code said this is only relevant for full-res raws
+  mCanonDoubleHeight &= Y_S_F == 1 && X_S_F == 1;
+  // old code said this is only relevant for s/mRaws
+  mCanonFlipDim &= X_S_F == 2;
 
-  //Prepare slices (for CR2)
-  uint32 slices = (uint32)slicesW.size() * (real_h - skipY) / 2;
-  offset = new uint32[slices+1];
-
-  uint32 t_y = 0;
-  uint32 t_x = 0;
-  uint32 t_s = 0;
-  uint32 slice = 0;
-  uint32 pitch_s = mRaw->pitch / 2;  // Pitch in shorts
-  slice_width = new int[slices];
-
-  // This is divided by comps, since comps pixels are processed at the time
-  for (uint32 i = 0 ; i <  slicesW.size(); i++)
-    slice_width[i] = slicesW[i] / COMPS;
-
-  for (slice = 0; slice < slices; slice++) {
-    offset[slice] = ((t_x + offX) * mRaw->getBpp() + ((offY + t_y) * mRaw->pitch)) | (t_s << 28);
-    _ASSERTE((offset[slice]&0x0fffffff) < mRaw->pitch*mRaw->dim.y);
-    t_y += 2;
-    if (t_y >= (real_h- skipY)) {
-      t_y = 0;
-      t_x += slice_width[t_s++];
-    }
-  }
-
-  // We check the final position. If bad slice sizes are given we risk writing outside the image
-  if ((offset[slices-1]&0x0fffffff) >= mRaw->pitch*mRaw->dim.y) {
-    ThrowRDE("LJpegPlain::decodeScanLeft: Last slice out of bounds");
+  if (mCanonDoubleHeight) {
+    mRaw->destroyData();
+    frame.h *= 2;
+    mRaw->dim = iPoint2D(frame.w * 2, frame.h);
+    mRaw->createData();
   }
 
-  offset[slices] = offset[slices-1];        // Extra offset to avoid branch in loop.
+  HuffmanTable* ht[N_COMP] = {0};
+  for (int i = 0; i < N_COMP; ++i)
+    ht[i] = huff[frame.compInfo[i].dcTblNo];
 
-  if (skipX)
-    slice_width[slicesW.size()-1] -= skipX;
+  mRaw->metadata.subsampling.x = X_S_F;
```

Comment:
You are right: it does not make sense here. It is there because it has been in the code before and I was approaching my cleanup with a "do what it did before but with less duplication" goal. I am still working on this part of the code where I started further looking into what parts actually make sens and what can be done in a simpler way. So this assignment will be removed, together with other unnecessary code.